Playoffs and Standings tiebreakers

jericksanjuan opened the pull request:
Based on:  http://stats.nba.com/playoffpicture/

Rank teams by Win Percentage (winp), Division Leaders (drank), Conference Win Percentage(cwinp), Win Percentage against other conference (ocwinp) and Point Differential (diff). 

Division Leaders are assured of a top 4 seeding per conference but will lose Home Court Advantage during playoffs if their matchup has the better winp to diff etc. 

Like in RL, this makes the division titles have more significance especially in the playoffs. And tiebreakers are really needed as it affects who makes the playoffs and also the matchups.

Missing from this change is the implementation of the following
(2) Head-to-head won-lost percentage
(4) W-L Percentage vs. Playoff teams, own conference
(5) W-L Percentage vs. Playoff teams, other conference

Because I'm not sure if adding them would  have a significant change in the results (other than the significant data store load and I think many usually clear game data like myself upon reaching many seasons played.)

dumbmatter commented on js/util/helpers.js:
```diff
@@ -882,6 +882,55 @@ define(["dao", "globals", "lib/knockout", "util/eventLog"], function (dao, g, ko
+     *
+     * @param {Array of String fields} sortBy sorting fields in order of application
+     */
+    function MultiSort(sortBy) {
```
Couldn't this be a normal function that returns the sortF function? If so, I'd prefer that to a constructor function.

dumbmatter commented on js/util/helpers.js:
```diff
@@ -882,6 +882,55 @@ define(["dao", "globals", "lib/knockout", "util/eventLog"], function (dao, g, ko
+     */
+    function MultiSort(sortBy) {
+
+        sortBy = sortBy;
```
Does nothing

dumbmatter commented on js/util/helpers.js:
```diff
@@ -882,6 +882,55 @@ define(["dao", "globals", "lib/knockout", "util/eventLog"], function (dao, g, ko
+     * Determine HCA advantage for this matchup. Team with better winp, cwinp,
+     * ocwinp and diff will get HCA regardless of seeding.
+     */
+    function seriesHomeAway(series, teamsConf, seed1, seed2, order, cid) {
```
Would be good to have unit tests for this, to make sure it's always working 100% for all edge cases.

dumbmatter commented:
`npm run lint` find a bunch of minor style problems.

jericksanjuan commented:
- Moved code from `newPhasePlayoffs` to `season.createPlayoffMatchups`. 
- Created tests for `createPlayoffMatchups` and `newSchedulePlayoffsDay`. This covers testing for proper playoff/standings order and determining hca.
- Changed `MultiSort` to just return anonymous function

dumbmatter commented on js/core/season.js:
```diff
@@ -697,6 +698,142 @@ define(["dao", "globals", "core/player", "core/team", "lib/bluebird", "lib/under
+                teams[i].won = 0;
+            }
+
+            if (!localStorage.top16playoffs || true) {
```
Don't want `|| true` here.

dumbmatter commented on js/core/team.js:
```diff
@@ -519,6 +533,53 @@ define(["dao", "globals", "core/player", "lib/bluebird", "lib/underscore", "util
+        };
+
+        if (options.sortBy.length > 0) {
+            assureSort('winp', null, ['won', 'lost']);
```
What's the point of checking all these things every time `filter` is called, and what happens if the requested sort is not possible?

jericksanjuan replied:
Optimized a bit. 

This is to make it more convenient for devs to get sorted teams. Fields not specified on the call to filter are added if needed by the sort field(s). Eliminates the need to remember adding fields like pts and oppPts if sorting by diff etc.

dumbmatter commented on js/util/helpers.js:
```diff
@@ -214,38 +219,247 @@ define(["dao", "globals", "lib/knockout", "util/eventLog"], function (dao, g, ko
     function getTeamsDefault() {
         var teams;
 
-        teams = [
```
I'm fine with most of your formatting changes, but can you revert this one? Makes the list of teams unwieldy to look at.

jericksanjuan replied:
Reverted

dumbmatter commented on js/util/helpers.js:
```diff
@@ -815,39 +1056,39 @@ define(["dao", "globals", "lib/knockout", "util/eventLog"], function (dao, g, ko
-                    contentNode.innerHTML = output;
-
-                    throw err;*/
+                    /*                    gSend = JSON.parse(JSON.stringify(g)); // deepCopy fails for some reason
```
Get rid of these added spaces

dumbmatter commented on js/test/core/season.js:
```diff
@@ -102,4 +106,109 @@ define(["globals", "core/season", "util/helpers", "test/helpers"], function (g,
+
+    describe('core/season playoffs tiebreakers', function() {
+        var tx;
+        before(function(done) {
```
`done` is not needed here, you can just return a promise and then it'll automatically realize the test is done when the promise resolves.

jericksanjuan replied:
Ah okay, missed that on the mocha doc.

dumbmatter commented:
> Division Leaders are assured of a top 4 seeding per conference but will lose Home Court Advantage during playoffs if their matchup has the better winp to diff etc. 

Can you get rid of that part? I always hated that rule.

jericksanjuan commented:
I just added the global option `divLeaderTop4`, if `true`, division leaders are ranked in the top 4 (default: false). I think having the option is better.
Also added helper function `getPlayoffSorting`, so that we can just set the default playoff/standings order in one place instead of per view that needs it.
